File: backend/app/services/complemento.py
"""El chat de una visita: completar lo que falto y preguntar sobre ella.

Existe porque una conversacion de campo nunca queda completa. El esposo que
maneja el cultivo llega a las tres de la tarde, el nombre de la finca quedo a
medias, el valor del arriendo se dijo sin decir si era mensual. Hasta ahora eso
moria en «temas pendientes» y solo se podia arreglar volviendo a la finca.

Aca el visitador lo escribe y el dato entra al registro.

La diferencia con la extraccion —y es la que gobierna este modulo— es de QUIEN
viene el dato. Alla habla el agricultor, grabado. Aca teclea el visitador, de
memoria y despues. Por eso todo lo que sale de este chat lleva
`fuente = Manual` y `hablante = visitador`, y eso no es una etiqueta: dispara
la primera regla dura del sistema —lo que dice el visitador nunca es
`Confirmado`— que la app y el backend aplican al escribir, cada uno por su
lado.

Ese marcado lo pone ESTE codigo, no el modelo. Al modelo se le pide que declare
lo que entendio; que sea un dato de segunda mano es un hecho del canal, no algo
sobre lo que tenga sentido preguntarle.
"""

import json

from anthropic import AsyncAnthropic
from fastapi import HTTPException

from ..config import Settings
from ..schemas_complemento import (
    MAX_MENSAJES,
    ComplementoRequest,
    ComplementoResult,
)
from .errors import UPSTREAM_EXCEPTIONS, upstream_error
from .extraccion import _lista_de_campos

SYSTEM = """Sos el asistente de UNA visita de campo de Sirius Regenerative.
Hablas con el VISITADOR que la hizo, no con el agricultor.

Servis para dos cosas, y las dos en la misma conversacion:

1. RESPONDER lo que el visitador pregunte sobre esta visita. Solo con lo que
   este en el contexto. Si el dato no esta registrado, decilo: "eso no quedo
   registrado". No lo completes con lo que suele pasar en una finca de la zona.

2. CAPTURAR lo que el visitador aporte. Cuando te cuenta algo que faltaba —"el
   esposo se llama Hernan", "el arriendo son 600 mil al mes", "el lote de
   platano tiene cuatro meses"— lo devolves como hallazgo contra el catalogo de
   campos.

COMO CAPTURAR:

- `clave`: exactamente una de las claves de la lista. Si lo que te cuenta no
  cabe en ninguna, NO inventes una clave: mandalo en `temas_pendientes`.
- `cita`: la frase del visitador, copiada TAL CUAL de su mensaje. Es la
  procedencia del dato: dentro de seis meses tiene que poder verse de donde
  salio un valor que no vino del audio.
- `certeza`:
  * "Confirmado": el visitador lo afirma sin dudar.
  * "Estimado": dio un aproximado, o dice que le parece.
  * "Inferido": lo deducis de lo que conto. Exige `razonamiento`.
  Escribi la que corresponda. El sistema aplica sus propias reglas encima y
  puede bajarla: un dato que aporta el visitador nunca vale lo mismo que uno
  que dijo el agricultor grabado, y eso no lo decidis vos.

- Si el visitador corrige un dato que ya estaba registrado, devolvelo igual con
  el valor nuevo. Que sea una correccion lo resuelve la app.

REGLAS QUE NO SE NEGOCIAN:

- No inventes datos. Si el visitador no lo dijo, no existe. Un campo vacio es
  un resultado correcto.
- No captures lo que el visitador pregunte, solo lo que afirme. "Cuanta tierra
  tiene?" es una pregunta, no un dato.
- No confirmes que guardaste algo que no devolviste como hallazgo. Si no cupo
  en ningun campo, decilo asi: "eso no tiene campo en el cuestionario, lo dejo
  anotado como pendiente".
- No prometas nada en nombre de Sirius: ni precios, ni productos, ni plazos.

COMO ESCRIBIS LA RESPUESTA:

En espanol, corto y directo. El visitador te lee de pie, con el telefono en una
mano. Cuando captures datos, decilos en una linea para que pueda ver si
entendiste bien — es su unica oportunidad de notar un error antes de que el
dato quede guardado. Texto plano, sin titulos ni tablas."""


def esquema(claves: list[str]) -> dict:
    """La herramienta que devuelve el modelo: respuesta mas hallazgos.

    `clave` es un enum cerrado, igual que en la extraccion: una clave inventada
    tiene que ser un error de validacion del proveedor y no un hallazgo que
    entra a la base apuntando a un campo que no existe.

    No lleva `segundo` ni `hablante`, al contrario de la extraccion: no hay
    audio de donde sacar un segundo, y el hablante es siempre el visitador —
    lo fija el codigo, no el modelo.
    """
    return {
        "type": "object",
        "properties": {
            "respuesta": {
                "type": "string",
                "description": "Lo que se le muestra al visitador en el chat.",
            },
            "hallazgos": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "clave": {"type": "string", "enum": claves},
                        "entidad_destino": {
                            "type": "string",
                            "enum": [
                                "Productor",
                                "Finca",
                                "Lote",
                                "Cultivo",
                                "Animal",
                                "Insumo",
                                "Visita",
                            ],
                        },
                        "entidad_local_id": {"type": ["string", "null"]},
                        "valor_texto": {"type": ["string", "null"]},
                        "valor_numerico": {"type": ["number", "null"]},
                        "unidad": {"type": ["string", "null"]},
                        "certeza": {
                            "type": "string",
                            "enum": ["Confirmado", "Estimado", "Inferido"],
                        },
                        "cita": {
                            "type": "string",
                            "description": (
                                "La frase del visitador, copiada tal cual."
                            ),
                        },
                        "razonamiento": {"type": ["string", "null"]},
                    },
                    "required": [
                        "clave",
                        "entidad_destino",
                        "certeza",
                        "cita",
                    ],
                    "additionalProperties": False,
                },
            },
            "temas_pendientes": {"type": "array", "items": {"type": "string"}},
        },
        "required": ["respuesta", "hallazgos", "temas_pendientes"],
        "additionalProperties": False,
    }


def _system(req: ComplementoRequest) -> str:
    partes = [SYSTEM]
    if req.visitador:
        partes.append(f"Estas hablando con {req.visitador}.")

    partes.append(
        "CAMPOS DEL CUESTIONARIO\n"
        "(Son los unicos donde se puede guardar un dato.)\n"
        + _lista_de_campos(req.campos)
    )

    if req.contexto:
        partes.append(
            "LO QUE YA TIENE ESTA VISITA\n"
            "(Es todo lo que sabes de ella. Lo que no este aca, no se "
            "registro.)\n" + req.contexto
        )
    else:
        partes.append(
            "ESTA VISITA NO TIENE NADA REGISTRADO TODAVIA. Decilo si te "
            "preguntan por ella; lo que el visitador aporte ahora es lo "
            "primero que va a quedar."
        )

    return "\n\n".join(partes)


def _marcar_procedencia(hallazgos: list[dict]) -> list[dict]:
    """Le pone a cada hallazgo de donde vino, y no se lo pregunta al modelo.

    `Manual` + `visitador` es lo que hace que la app y Airtable apliquen la
    regla dura al escribir. Si el modelo pudiera elegir estos dos campos,
    bastaria una alucinacion para que un dato tecleado por el visitador
    entrara al registro como si lo hubiera afirmado el agricultor frente a una
    grabadora.
    """
    marcados = []
    for h in hallazgos:
        if not isinstance(h, dict) or not h.get("clave"):
            continue
        marcados.append(
            {
                **h,
                "fuente": "Manual",
                "hablante": "visitador",
                # No hay audio detras de esto. Un segundo inventado mandaria al
                # visitador a escuchar un minuto donde nadie dijo nada.
                "segundo": None,
            }
        )
    return marcados


async def responder(
    settings: Settings, req: ComplementoRequest
) -> ComplementoResult:
    if not settings.anthropic_api_key:
        raise HTTPException(
            status_code=500, detail="Falta ANTHROPIC_API_KEY en el backend."
        )
    if not req.campos:
        raise HTTPException(
            status_code=400,
            detail="No llego el catalogo de campos, asi que no se puede armar el esquema.",
        )

    mensajes = [
        {"role": m.rol, "content": m.contenido}
        for m in req.mensajes[-MAX_MENSAJES:]
        if m.contenido.strip()
    ]
    if not mensajes or mensajes[0]["role"] != "user":
        raise HTTPException(
            status_code=400,
            detail="La conversacion empieza con un mensaje del visitador.",
        )

    claves = [c.clave_tecnica for c in req.campos]
    herramienta = {
        "name": "complementar_visita",
        "description": (
            "Responde al visitador y devuelve los datos que aporto en su "
            "mensaje."
        ),
        "input_schema": esquema(claves),
    }

    cliente = AsyncAnthropic(api_key=settings.anthropic_api_key)
    try:
        respuesta = await cliente.messages.create(
            model=settings.claude_model,
            max_tokens=2000,
            system=_system(req),
            messages=mensajes,
            tools=[herramienta],
            tool_choice={"type": "tool", "name": "complementar_visita"},
        )
    except UPSTREAM_EXCEPTIONS as exc:
        raise upstream_error("Anthropic", exc) from exc

    datos = None
    for bloque in respuesta.content:
        if bloque.type == "tool_use":
            datos = bloque.input
            break

    if datos is None:
        raise HTTPException(
            status_code=502,
            detail="El modelo no devolvio la respuesta en el formato esperado.",
        )

    if isinstance(datos, str):
        datos = json.loads(datos)

    texto = (datos.get("respuesta") or "").strip()
    if not texto:
        raise HTTPException(
            status_code=502, detail="El modelo devolvio una respuesta vacia."
        )

    return ComplementoResult(
        respuesta=texto,
        hallazgos=_marcar_procedencia(datos.get("hallazgos") or []),
        temas_pendientes=[
            t for t in (datos.get("temas_pendientes") or []) if str(t).strip()
        ],
        modelo=settings.claude_model,
    )

File: app/lib/state/complemento.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../core/api_client.dart';
import '../core/complemento_visita.dart';
import '../data/visita_repository.dart';
import 'providers.dart';
import 'sesion.dart';

class ComplementoState {
  const ComplementoState({
    this.mensajes = const [],
    this.esperando = false,
    this.error,
    this.escritos = const [],
    this.ultimoTurnoEscribio = 0,
  });

  final List<MensajeChat> mensajes;

  /// Hay una pregunta en vuelo. Bloquea mandar otra: el hilo tiene que
  /// alternar visitador/modelo o el backend lo rechaza.
  final bool esperando;

  final String? error;

  /// Los datos que esta conversacion ya dejo escritos, en texto legible. Se
  /// acumulan para poder decir al final que se registro, y viajan al markdown
  /// que se guarda en Airtable.
  final List<String> escritos;

  /// Cuantos datos escribio el ULTIMO turno. La pantalla lo usa para avisarlo
  /// en el momento: el dato entra directo al registro, y ese es el unico
  /// instante en que alguien puede notar que el modelo entendio mal.
  final int ultimoTurnoEscribio;
}

/// El chat de una visita. Responde sobre ella y ESCRIBE lo que el visitador
/// aporte.
///
/// Es el unico camino del sistema donde un dato entra al registro sin haber
/// pasado por el audio. Tres cosas lo hacen aceptable, y ninguna esta en el
/// prompt:
///
/// 1. El backend marca cada hallazgo `fuente = Manual`, `hablante = visitador`.
/// 2. La app los guarda por `guardarExtraccion`, que aplica las reglas duras:
///    lo dicho por el visitador nunca queda `Confirmado`.
/// 3. Cada dato lleva como cita la frase con la que el visitador lo dijo, asi
///    que meses despues se puede ver de donde salio.
///
/// La conversacion entera se guarda con la visita y sube a Airtable: es el
/// registro de procedencia de todo lo anterior.
class ComplementoController extends StateNotifier<ComplementoState> {
  ComplementoController(this._ref, this.visitaId)
      : super(const ComplementoState());

  final Ref _ref;
  final String visitaId;

  Future<void> enviar(String texto) async {
    final mensaje = texto.trim();
    if (mensaje.isEmpty || state.esperando) return;

    final hilo = [
      ...state.mensajes,
      MensajeChat(rol: 'user', contenido: mensaje),
    ];
    state = ComplementoState(
      mensajes: hilo,
      esperando: true,
      escritos: state.escritos,
    );

    final repo = _ref.read(repoProvider);
    try {
      // El contexto se relee en cada turno: el visitador puede haber
      // complementado tres datos hace un minuto, y el modelo no puede volver a
      // pedirlos como si faltaran.
      final contexto = await repo.contextoComplemento(visitaId);
      final campos = await repo.camposParaExtraccion();
      final sesion = _ref.read(sesionProvider).valueOrNull;
      final visitador = sesion?.credencial.nombre;

      final resultado = await _ref.read(apiProvider).complementar(
            codigoVisita: visitaId,
            mensajes: hilo,
            campos: campos,
            contexto: contexto,
            visitador: visitador,
          );

      final completo = [
        ...hilo,
        MensajeChat(rol: 'assistant', contenido: resultado.respuesta),
      ];

      // Se guarda lo que el turno aporto ANTES de responder en la pantalla: si
      // la app muriera entre una cosa y la otra, es mejor que el dato este
      // guardado y el visitador lo vuelva a leer, que verlo confirmado en una
      // burbuja y que no exista.
      final escritos = await _guardar(resultado, repo);

      state = ComplementoState(
        mensajes: completo,
        escritos: [...state.escritos, ...escritos],
        ultimoTurnoEscribio: escritos.length,
      );

      await _archivar(completo, visitador);
    } on ApiException catch (e) {
      state = ComplementoState(
        mensajes: hilo,
        error: e.message,
        escritos: state.escritos,
      );
    } catch (e) {
      state = ComplementoState(
        mensajes: hilo,
        error: '$e',
        escritos: state.escritos,
      );
    }
  }

  /// Escribe los hallazgos del turno y devuelve como quedaron, en texto.
  ///
  /// Va por `guardarExtraccion` —el mismo camino del audio— y no por una
  /// insercion propia: ahi viven las reglas duras y el recalculo de la
  /// completitud. Un camino paralelo seria un camino sin reglas.
  Future<List<String>> _guardar(
    ComplementoResult resultado,
    VisitaRepository repo,
  ) async {
    if (resultado.hallazgos.isEmpty && resultado.temasPendientes.isEmpty) {
      return const [];
    }

    final extraidos = [
      for (final h in resultado.hallazgos) HallazgoExtraido.fromJson(h),
    ];

    await repo.guardarExtraccion(
      visitaId: visitaId,
      extraidos: extraidos,
      // El resumen NO se toca: el de la visita lo escribio el modelo con la
      // conversacion entera delante, y este chat solo vio un mensaje suelto.
      temasPendientes: resultado.temasPendientes,
    );

    return [
      for (final h in extraidos)
        '${h.claveTecnica}: '
            '${h.valorTexto ?? h.valorNumerico ?? ''}'
            '${h.unidad == null ? '' : ' ${h.unidad}'}',
    ];
  }

  /// Guarda la conversacion con la visita y la encola para Airtable.
  ///
  /// No tumba el turno si falla: el dato ya quedo escrito y la respuesta ya
  /// esta en la pantalla. Que el registro de la conversacion no se guarde es
  /// una perdida menor comparada con mostrarle un error al visitador por algo
  /// que ya funciono.
  Future<void> _archivar(List<MensajeChat> hilo, String? visitador) async {
    try {
      await _ref.read(repoProvider).guardarConversacionComplemento(
            visitaId: visitaId,
            contenido: markdownConversacion(
              mensajes: hilo,
              generadoEn: DateTime.now(),
              visitador: visitador,
              datosEscritos: state.escritos,
            ),
          );
    } catch (_) {
      // Silencioso a proposito, ver arriba.
    }
  }

  void limpiarError() => state = ComplementoState(
        mensajes: state.mensajes,
        escritos: state.escritos,
      );
}

final complementoProvider = StateNotifierProvider.family<ComplementoController,
    ComplementoState, String>(
  (ref, visitaId) => ComplementoController(ref, visitaId),
);
